File: 1000/D_Coprime.cpp
#include <bits/stdc++.h>

using namespace std;

void srvkr() {
    int n; cin >> n;
    vector<int> v(n);
    for (auto & i : v) {
        cin >> i;
    }
    int ans = -1;
    unordered_set<int> us;
    for (int i = n - 1; i >= 1; --i) {
        if (us.find(v[i]) != us.end()) {
            continue;
        } else {
            us.insert(v[i]);
        }
        int j = i;
        while  (j >= 0 && __gcd(v[i], v[j]) != 1) {
            --j;
        }
        if (j >= 0) {
            if (ans < (i + j + 2)) {
                ans = i + j + 2;
            }
        }
    }
    cout << ans << '\n';
}

int32_t main() {
    int tc;
    cin >> tc;
    while (tc--) {
        srvkr();
    }
    return 0;
}

File: 1000/D_Coprime.java
import java.util.HashSet;
import java.util.Scanner;

public class D_Coprime {
    static Scanner cin = new  Scanner(System.in);
    public static void main(String[] args) {
        int tc = 1;
        tc = cin.nextInt();
        while (tc > 0) {srvkr(); --tc;}
    }

    public static void srvkr() {
        int n = cin.nextInt(); int[] input = new int[n];
        for (int i = 0; i < n; ++i) {
            input[i] = cin.nextInt();
        }
        int ans = -1;
        HashSet<Integer> hs = new HashSet<>();
        for (int i = n - 1; i >= 1; --i) {
            if (hs.contains(input[i])) {
                continue;
            } else {
                hs.add(input[i]);
            }
            int j = i;
            while (j >= 0 && gcd(input[i], input[j]) != 1) {
                --j;
            }
            if (j >= 0) {
                if (ans < (i + j + 2)) {
                    ans = i + j + 2;
                }
            }
        }
        System.out.println(ans);
    }
    public static int gcd(int a, int b) {
        int r = a % b;
        while (r != 0) {
            a = b;
            b = r;
            r = a % b;
        }
        return b;
    }
}